// pdf-lib/lib/core/pdf-structures/index.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
}
Object.defineProperty(exports, "__esModule", { value: true });
var PDFXRef = __importStar(require("./PDFXRef"));
exports.PDFXRef = PDFXRef;
var PDFCatalog_1 = require("./PDFCatalog");
exports.PDFCatalog = PDFCatalog_1.default;
var PDFContentStream_1 = require("./PDFContentStream");
exports.PDFContentStream = PDFContentStream_1.default;
var PDFObjectStream_1 = require("./PDFObjectStream");
exports.PDFObjectStream = PDFObjectStream_1.default;
var PDFHeader_1 = require("./PDFHeader");
exports.PDFHeader = PDFHeader_1.default;
var PDFPage_1 = require("./PDFPage");
exports.PDFPage = PDFPage_1.default;
var PDFPageTree_1 = require("./PDFPageTree");
exports.PDFPageTree = PDFPageTree_1.default;
var PDFTrailer_1 = require("./PDFTrailer");
exports.PDFTrailer = PDFTrailer_1.default;
var PDFLinearizationParams_1 = require("./PDFLinearizationParams");
exports.PDFLinearizationParams = PDFLinearizationParams_1.default;
var PDFXRefStream_1 = require("./PDFXRefStream");
exports.PDFXRefStream = PDFXRefStream_1.default;
